sound/soc/intel/boards/bdw-rt5677.c

```diff
@@ -423,8 +423,22 @@ static int bdw_rt5677_probe(struct platform_device *pdev)
 	return devm_snd_soc_register_card(&pdev->dev, &bdw_rt5677_card);
 }
 
+static int bdw_rt5677_remove(struct platform_device *pdev)
+{
+	struct bdw_rt5677_priv *bdw_rt5677;
+
+	bdw_rt5677 = snd_soc_card_get_drvdata(&bdw_rt5677_card);
+
+	if (bdw_rt5677->component)
+		devm_gpiod_put(bdw_rt5677->component->dev,
```

Comment:
@plbossart I think this is partly because of the order in which we're remove the modules. The machine drv is removed first but the devm_gpio_get() is associated with the codec component dev. I have an odd suggestion, not sure if it will work but can we do:
```
bdw_rt5677->gpio_hp_en = devm_gpiod_get(rtd->card->dev, "headphone-enable",
						GPIOD_OUT_LOW);

```

Reply:
Looking at the code I think the following should be done: a) introduction of ->exit() callback for DAI link, b) implementation of it in this code, c) due to above drop devm_*() there in a favor of plain ones.

Reply:
Looking at the code I think the following should be done: a) introduction of ->exit() callback for DAI link, b) implementation of it in this code, c) due to above drop devm_*() there in a favor of plain ones.

Reply:
I tried that and got an error, could not get the gpio at all.

Reply:
this is what I get with the device change

````
[   28.145202] rt5677 i2c-RT5677CE:00: Can't find HP_AMP_SHDN_L gpio
[   28.145205] bdw-rt5677 bdw-rt5677: ASoC: failed to init Codec: -2
````

Reply:
> @plbossart
> Looking at the code I think the following should be done: a) introduction of ->exit() callback for DAI link, b) implementation of it in this code, c) due to above drop devm_*() there in a favor of plain ones.

Sounds indeed like a good idea. We should always have dual functions. I just tested this and it seems to work fine.